File: hs/Math/Numerical/LinearSystem.hs
{-# LANGUAGE CPP #-}
module Math.Numerical.LinearSystem
	(
	  solveConjGrad
	, solveConjGradGeneric
	) where

import qualified Data.List as List
import qualified Data.Array as Array
import qualified System.IO.Unsafe as IOUnsafe

import Foreign
import Foreign.C.Types


#if x86_64_HOST_ARCH
foreign import ccall "_ZN4Math9Numerical22conjugateGradientSolveEmmPKmPKdS4_Pd"
#else
foreign import ccall "_ZN4Math9Numerical22conjugateGradientSolveEjjPKjPKdS4_Pd"
#endif
	c_conjugateGradientSolve :: CSize -> CSize -> Ptr CSize -> Ptr CDouble -> Ptr CDouble -> Ptr CDouble -> IO CDouble

--foreign import ccall "_ZN4Math9Numerical21gaussEliminationSolveEjPPdS1_S1_"
--	c_gaussEliminationSolve :: CSize -> Ptr (Ptr CDouble) -> Ptr CDouble -> Ptr CDouble -> IO CDouble


checkAndSimplify :: Int -> [((Int, Int), Double)] -> [(Int, Double)] -> ([[(Int, Double)]], [Double])
checkAndSimplify n rawA rawB
	| n <= 0     = error "solveConjGrad: n must be positive"
	| not okA    = error "solveConjGrad: matrix indices out of bound"
	| not okB    = error "solveConjGrad: vector indices out of bound"
	| otherwise  = (aList, bList)

	where
		okA = List.all (\ ((i, j), _) -> i >= 0 && i < n && j >= 0 && j < n) rawA
		okB = List.all (\ (i, _) -> i >= 0 && i < n) rawB

		aList = snd $ List.mapAccumL accum reducedA [0 .. (n - 1)]
			where
				reducedA =
					let grouped = List.groupBy (\ (i, _) (j, _) -> i == j) $ List.sort rawA
					in map (foldl1 (\ (i, a) (_, b) -> (i, a + b))) grouped 

				accum list i =
					let (cur, rest) = List.span (\ ((ci, _), _) -> ci == i) list
					in (rest, map (\ ((_, j), a) -> (j, a)) cur)

		bList = snd $ List.mapAccumL accum (List.sort rawB) [0 .. (n - 1)]
			where
				accum list i =
					let (cur, rest) = List.span (\ (ci, _) -> ci == i) list
					in (rest, List.sum $ snd $ unzip cur)


solveConjGrad :: Int -> [((Int, Int), Double)] -> [(Int, Double)] -> [Double]
solveConjGrad n rawA rawB =
	IOUnsafe.unsafePerformIO $ do
		x <- mallocArray n :: IO (Ptr CDouble)

		_ <- withArray (map realToFrac bList) $ \ b ->
			withArray (map fromIntegral nid) $ \ ids ->
				withArray (map realToFrac na) $ \ a ->
					c_conjugateGradientSolve (fromIntegral n) (fromIntegral m) ids a b x

		solution <- fmap (map realToFrac) $ peekArray n x
		free x
		return solution

	where
		(aList, bList) = checkAndSimplify n rawA rawB

		(na, nid) = unpack (List.concat $ map (\ (i, lst) -> map (\ (j, a) -> ((i, j), a)) lst) $ zip [(0 :: Int) ..] aList) ([], [])
			where
				unpack [] res = res
				unpack (((i, j), a) : rest) (ra, rid) = unpack rest (a : ra, i : j : rid)

		m = length na


solveConjGradGeneric :: Int -> [((Int, Int), Double)] -> [(Int, Double)] -> [Double]
solveConjGradGeneric n rawA rawB = conjGrad (0, initialX, initialR, initialR, delta0)
	where
		(aList, bList) = checkAndSimplify n rawA rawB

		evaluate xl = map (List.sum . map (\ (i, a) -> a * (x Array.! i))) aList
			where
				x = Array.listArray (0, n - 1) xl

		difference x = zipWith (-) bList $ evaluate x

		dot la lb = List.sum $ zipWith (*) la lb

		initialX = take n $ List.repeat 0.0
		initialR = difference initialX
		delta0 = dot initialR initialR

		updateState (k, x, d, r, delta) = (k + 1, newX, newD, newR, newDelta)
			where
				q = evaluate d
				alpha = delta / (dot d q)

				newX = zipWith (\ xi di -> xi + alpha * di) x d

				newR = if 0 == mod k 50
					then difference newX
					else zipWith (\ ri qi -> ri - alpha * qi) r q

				newDelta = dot newR newR

				newD = let beta = newDelta / delta
					in zipWith (\ ri di -> ri + beta * di) newR d

		eps = 1e-15

		conjGrad st@(k, x, _, _, delta) =
			if k > n || delta <= delta0 * eps * eps
				then x
				else conjGrad $ updateState st

File: hs/Math/Numerical/LinearSystem.cpp
#include <cstddef>
#include <cmath>
#include <algorithm>

namespace Math { namespace Numerical {

	inline double dot(const size_t n, const double * x, const double * y)
	{
		double dot = 0.0;
		for(size_t i = 0; i < n; i++)
			dot += x[i] * y[i];

		return dot;
	}

	inline void multiply(const size_t n
		, const size_t m
		, const size_t * id
		, const double * a
		, const double * x
		, double * r)
	{
		for(size_t i = 0; i < n; i++)
			r[i] = 0.0;

		for(size_t i = 0; i < m; i++)
		{
			const size_t p = id[2 * i];
			const size_t q = id[2 * i + 1];
			r[p] += a[i] * x[q];
		}
	}

	inline void diff(const size_t n
		, const size_t m
		, const size_t * id
		, const double * a
		, const double * b
		, const double * x
		, double * r)
	{
		multiply(n, m, id, a, x, r);
		for(size_t i = 0; i < n; i++)
			r[i] = b[i] - r[i];
	}

	double conjugateGradientSolve(const size_t n
		, const size_t m
		, const size_t * id
		, const double * a
		, const double * b
		, double * x)
	{
		for(size_t i = 0; i < n; i++)
			x[i] = 0.0;

		double * r = new double[n];
		diff(n, m, id, a, b, x, r);

		double * d = new double[n];
		for(size_t i = 0; i < n; i++)
			d[i] = r[i];

		double * q = new double[n];

		double delta_new = dot(n, r, r);
		const double delta0 = delta_new;
		const double eps = 1e-14;

		for(size_t k = 0; k < n && delta_new > eps * eps * delta0; k++)
		{
			multiply(n, m, id, a, d, q);
			double alpha = delta_new / dot(n, d, q);

			for(size_t i = 0; i < n; i++)
				x[i] += alpha * d[i];

			if(k % 50 == 0)
				diff(n, m, id, a, b, x, r);
			else
			{
				for(size_t i = 0; i < n; i++)
					r[i] -= alpha * q[i];
			}

			const double delta_old = delta_new;
			delta_new = dot(n, r, r);

			{
				const double beta = delta_new / delta_old;
				for(size_t i = 0; i < n; i++)
					d[i] = r[i] + beta * d[i];
			}
		}

		delete[] r;
		delete[] d;
		delete[] q; 

		return delta_new;
	}

	double gaussEliminationSolve(const size_t n
		, double ** a
		, double * b
		, double * x)
	{
		for(size_t k = 0; k < n; k++)
		{
			size_t best = k;
			for(size_t i = k + 1; i < n; i++)
				if(fabs(a[i] [k]) > fabs(a[best] [k]))
					best = i;

			if(best != k)
			{
				std::swap(a[best], a[k]);
				std::swap(b[best], b[k]);
			}

			const double * leadRow = a[k];
			const double leadElement = a[k] [k];

			for(size_t i = k + 1; i < n; i++)
			{
				const double coef = a[i] [k] / leadElement;
				for(size_t j = k; j < n; j++)
					a[i] [j] -= coef * leadRow[j];
				b[i] -= coef * b[k];
			}
		}

		for(size_t _ = 0; _ < n; _++)
		{
			const size_t i = n - 1 - _;

			double cur = b[i];
			for(size_t j = i + 1; j < n; j++)
				cur -= x[j] * a[i] [j];

			x[i] = cur / a[i] [i];
		}

		return 0.0;
	}

}}
